scenes/cslib_array.txt

```diff
@@ -247,6 +247,180 @@
 *return
 
 
+*comment TO_STRING
+*comment ------------------------------
+*comment Returns all elements of an array joined as a string.
+*comment An optional string may be passed with which
+*comment to separate the elements.
+*comment ------------------------------
+*comment    params:
+*comment        p_array_ref(string): the name of the array to convert to a string
+*comment        p_separator(string)?: the string to insert between each array element
+*comment    returns:
+*comment        (any): all array elements as a string
+*label to_string
+*params p_array_ref
+*temp p_separator "@{(param_count > 1) ${param_2}|}"
+*temp count {p_array_ref&"_count"}
+*set cslib_ret ""
+*if (count < 1)
+    *return
+*temp n 1
+*label _to_string_loop
+*if (n < count)
+    *set cslib_ret &({p_array_ref&"_${n}"}&p_separator)
+    *set n + 1
+    *goto _to_string_loop
+*set cslib_ret &{p_array_ref&"_${n}"}
+*return
+
+
+*comment REVERSE
+*comment ------------------------------
+*comment Reverses the elements in the array.
+*comment ------------------------------
+*comment    params:
+*comment        p_array_ref(string): the name of the array to reverse
+*label reverse
+*params p_array_ref
+*set cslib_ret ""
+*temp count {p_array_ref&"_count"}
+*if count < 2
+    *return
+*temp s_idx 1
+*temp e_idx count
+*temp store
+*label _reverse_loop
+*if (s_idx < e_idx)
+    *set store {p_array_ref&"_${s_idx}"}
+    *set {p_array_ref&"_${s_idx}"} {p_array_ref&"_${e_idx}"}
+    *set {p_array_ref&"_${e_idx}"} store
+    *set s_idx + 1
+    *set e_idx - 1
+    *goto _reverse_loop
+*return
+
+
+*comment FOR_EACH
+*comment ------------------------------
+*comment Call a given subroutine against each
+*comment element in the array.
+*comment
+*comment The given routine should expect
+*comment to be passed three parameters:
+*comment - The element value
+*comment - The element's source array index
+*comment - The name of the array
+*comment
+*comment And should return a new value for the element.
+*comment E.g.:
+*comment    *label capitalize_str
+*comment    *params value index array_ref
+*comment    *set cslib_ret "$!!{value}"
+*comment    *return
+*comment ------------------------------
+*comment    params:
+*comment        p_array_ref(string): the name of the array to iterate over
+*comment        p_scene(string): the name of the scene containing the routine
+*comment        p_label(string): the name of the label denoting the start of the routine
+*label for_each
+*params p_array_ref p_scene p_label
+*set cslib_ret ""
+*temp count {p_array_ref&"_count"}
+*temp n 1
+*label _for_each_loop
+*if (n <= count)
+    *gosub_scene {p_scene} {p_label} {p_array_ref&"_${n}"} n p_array_ref
```

Comment:
wait, why do you need to pass
array[n]   n   array
can't the function determine array[n] out of array and n?

Reply:
Yes, but you've got multiple comments on this patch saying how ugly doing that is... ;)
Best we keep that away from the end-suer, so this is a convenience value (it doesn't cost us anything).